HighscoreWindow.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2022, Harshit Sharma <harshits908@gmail.com>
+ * Copyright 2023, Johan Wagenheim
+ * This file is distributed under the terms of the MIT license
+ */
+
+#include "HighscoreWindow.h"
+#include "WindowBoard.h"
+#include "Game.h"
+
+#include <Catalog.h>
+#include <string>
+#include <LayoutBuilder.h>
+#include <Message.h>
+#include <Messenger.h>
+#include <Button.h>
+#include <String.h>
+#include <StringView.h>
+
+#undef B_TRANSLATION_CONTEXT
+#define B_TRANSLATION_CONTEXT "HighscoreWindow"
+
+HighscoreWindow::HighscoreWindow(const char* oldHighscorer, const int32 oldHighscore, 
+	const int32 newHighscore, const BMessenger &messenger)
+	:
+	BWindow(BRect(200, 200, 1,1), 0, B_MODAL_WINDOW, B_NOT_CLOSABLE | B_NOT_RESIZABLE),
+				fMessenger(messenger)
+{
+	BStringView* congratulations = new BStringView("congratulations",
+											B_TRANSLATE("Congratulations!"));
+	congratulations->SetFont(be_bold_font);
+	congratulations->SetFontSize(25);
+	congratulations->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	
+	BString newHighscoreText = B_TRANSLATE("New highscore: %newhighscore%");
+	newHighscoreText.ReplaceFirst("%newhighscore%", std::to_string(newHighscore).c_str());
+	BStringView* newHighScore = new BStringView("newHighScore", newHighscoreText.String());
+	newHighScore->SetFont(be_bold_font);
+	newHighScore->SetFontSize(20);
+	newHighScore->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	
+	BStringView* inputlabel = new BStringView("inputlabel",
+										B_TRANSLATE("Please enter your name:"));
+	inputlabel->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	inputlabel->SetFont(be_bold_font);
+
+	fInputBox = new
+		BTextControl("NameInput", NULL, NULL, new BMessage(H2048_NAME_ENTERED));
+	fInputBox->SetText(B_TRANSLATE("Anonymous"));
+	fInputBox->SetAlignment(B_ALIGN_HORIZONTAL_CENTER,B_ALIGN_HORIZONTAL_CENTER);
+	fInputBox->MakeFocus();
+	
+	BString previousWinner;
+	previousWinner << B_TRANSLATE("This is the first record!");
+	if (oldHighscorer[0]) {
+		// Replace with previous winner and score difference
+		previousWinner = B_TRANSLATE("You've beaten the previous record holder\n"
```

Comment:
I think the %oldHighscorer should be between commas:
```
"You've beaten the previous record holder,\n"
"%oldHighscorer%, by %deltaScore%."
```

Or maybe even simpler:
```
"You've beaten "%oldHighscorer%'s\n"
"old record by %deltaScore%."
```

Your call.

Reply:
OK, I updated with the first option (and squashed the commits).

I'm unsure how to limit the input length, any hints?

Reply:
> I'm unsure how to limit the input length, any hints?

Either, you silently truncate the string when the user is done entering.
Or, you can use the BTextControl's SetModificationMessage() that sends a message for every letter entered. Then check with every letter if the limit was reached and remove the last entered letter. That should prevent the user to enter any more letters. I think.
The second option should be enough to signal the input limit to the user. I wouldn't pop up an alert or something. It's obvious enough.

Reply:
Thanks, I added a SetModificationMessage() to make sure the input is limited to 25 characters, and applied some other recommendations from @korli . Please review when you have a chance.